src/fmu/dataio/providers/objectdata/_tables.py

```diff
@@ -23,39 +24,109 @@
+
+    table_index = [col for col in standard_index.columns if col in table_columns]
+    if not table_index:
+        warnings.warn(
+            "Could not detect any standard index columns in table: "
+            f"{standard_index.columns}. If the table has index columns they "
+            "should be provided as input through the 'table_index' argument."
+        )
     else:
-        index = table_index
+        if set(standard_index.required) != set(table_index):
+            warnings.warn(
+                "The table provided does not contain all required standard "
+                f"table index columns for this content: {standard_index.required}. "
+                "This may not be allowed in the future.",
+                FutureWarning,
+            )
+    return table_index
+
+
+def _derive_index_legacy(table_columns: list[str]) -> list[str]:
+    """
+    Derive all columns in table that is registered as a standard index column,
+    independent of the content.
+    """
+    table_index = []
+    for standard_table_index in STANDARD_TABLE_INDEX_COLUMNS.values():
+        for col in standard_table_index.columns:
+            if col in table_columns and col not in table_index:
+                table_index.append(col)
+    return table_index
+
+
+def _derive_index(
+    table_columns: list[str],
+    table_index: list[str] | None,
+    content: Content | None = None,
+) -> list[str]:
+    """
+    Derive the index for a table based on a provided table_index or using
+    standard index columns for the specific content.
+
+    If no table index is provided and content is not registered with any standard table
+    indexes, the index is set using the legacy method; columns in table that
+    matches any registered standard index columns (independent of content).
+    """
+    if table_index:
+        logger.debug("Table index provided, validating input...")
+        _validate_input_table_index(table_index, table_columns, content)
+        return table_index
 
-    if "REAL" in columns:
-        index.append("REAL")
-    _check_index_in_columns(index, columns)
-    return index
+    logger.debug("No table index provided")
+
+    if content in STANDARD_TABLE_INDEX_COLUMNS:
+        table_index = _derive_index_from_standard(content, table_columns)
+    else:
+        table_index = _derive_index_legacy(table_columns)
+        if table_index:
+            warnings.warn(
+                "The 'table_index' was not provided, and has been set based upon "
+                "commonly known table index columns found in the table: "
+                f"{table_index}. In the future the table_index will be empty in the "
+                "metadata if not provided as input through the 'table_index' argument.",
+                FutureWarning,
+            )
+    # TODO: Look into removing this, probably only applicable for aggregated data.
+    if "REAL" in table_columns and "REAL" not in table_index:
+        table_index.append("REAL")
```

Comment:
I was unsure about keeping this, I think it is only relevant to aggregated data. And going forward we want users to use the `fmu-sumo-aggregation-service` to do aggregations, and not uploading aggregated data directly. 
The aggregation service has it's own code for adding `REAL` column to the tables and adding it to the `table_index` in the metadata.

I landed on keeping it for now, as it do no harm.. what do you think @mferrera?